## A Hands-on Workshop series in Machine Learning
### Session 1: Data Manipulation using pandas

### Meet the Instructor:
Aashita Kesarwani  
Current job: Scientific Computing  and Data Science Specialist at Harvey Mudd College  
Background:  
- PhD in Mathematics from Tulane University (Number Theory)
- Undergraduate from IIT (Indian Institute of Technology) in Applied Mathematics    
- Visiting AI Researcher at [deepkapha.ai](https://www.linkedin.com/company/digitalis-kapha-b-v-/) in 2019
- [Open source contributor](https://pypi.org/user/Aashita/)

Open your Anaconda Navigator and then open the Jupyter notebook.  
Create a folder for the workshop and save the Session 1 folder inside it.  
Navigate to this folder in your laptop from your Jupyter web browser.  
Open this notebook titled "*Data Manipulation using pandas.ipynb*"

Please:
- Resize the Zoom and Web browser with Jupyter notebook so that you have both of them side by side. 
- Ask questions. Type in the chat window or raise your hand to speak.
- Answer questions and participate in the discussion - type in the chat window or unmute yourself to speak.
- Ask for help, whenever needed, for the exercises.
- Mute yourself when not speaking. 

If the Jupyter notebook or Python is not running in your laptop or you do not have all the required packages installed, you can use [Google Colab](https://colab.research.google.com/). Colab runs on the cloud in a web browser on Google machines, so it does not depend on Python installation in your laptop.

##### What is machine learning?
- Learning from data without explicit programming.

##### How to approach a problem? 
- Data exploration and feature engineering 
- Model building, tuning and testing 
    
#### Topics to be covered today (Data manipulation using `pandas`)
- A brief overview of Jupyter Notebook
- Pandas dataframes as the data structure for datasets
- Slicing dataframes using conditionals as well as iloc and loc methods.
- Statistical summary and exploration of dataframes
- Basic plots
- Feature engineering:
    - Detecting and filling missing values in the dataframes
    - Creating new features; using regular expressions for data extraction
    - Dropping rows/columns; replacing values of a column using a dictionary
- Merging columns from different datasets (Optional)

Today's session is meant to ease you into the very first step of the machine learning project - cleaning and exploring the data. The machine learning algorithms will be introduced from the next session, when we will start building models for classification tasks.

##### Structure of the session:

The notebook is divided into the following sections:  

0. Preliminaries  
1. Slicing rows and columns from the dataframe
2. Exploring the dataset (45 min exercise session)
3. Feature Engineering: Creating a new column for the titles of the passengers (30 min exercise session)
4. Merging columns from different datasets (Optional) 

You will follow along with me in some of the sections and the rest are the exercises that you will work on in groups in Zoom breakout rooms.

##### Note:
* Solutions for all the exercises (including the ones in the follow-along section) will be uploaded in the same folder of the [Github repository](https://github.com/AashitaK/A-Hands-on-Workshop-series-in-Machine-Learning). There will also be a link for the lecture capture in the [webpage](http://www.aashitak.com/A-Hands-on-Workshop-series-in-Machine-Learning/).
* Please give your feedback (form will shared with you during the workshop) and it will be gladly taken into account for the next sessions.

### 0. Preliminaries
#### An overview of ***Jupyter Notebook***.
Jupyter Notebook is the de facto standard in Data Science inspired by the concept of [literate programming](https://www-cs-faculty.stanford.edu/~knuth/lp.html) introduced by [Donald Knuth](https://amturing.acm.org/award_winners/knuth_1013846.cfm). 

It is composed of blocks that are called cells.

There are two types of cells:
* Code cells
* Markdown cells (for text such as this cell itself)

You can run the code within the cells (No need to use command line) by first selecting the cell and then using `Shift` + `Enter`. Another option is to use the `Run` button at the header at the top.

In [2]:
a = 2
b = 3
a + b

5

#### Key shortcuts using command mode
Press `Esc` to activate the command mode:
Shortcuts:
* A: Insert cell above
* B: Insert cell below
* C: Copy 
* V: Paste 
* X: Cut 
* DD: Delete 
* M: Convert a cell to Markdown cell
* Shift: Let's you select multiple cells at once that you can copy/cut/delete.

To exit the command mode, simply press `Enter`. You need to first exit the command mode to run/edit the cell. 

Jupyter notebook allows Tab completion, so as to let you avoid typing the function and variable names. 

It also supports the help feature. 
- To get documentation about a function, you pass the name of the function inside `help()`.
- To view help, you can also type the function name followed by `?`.
- To get the source code of a function, type the function followed by a double question mark `??`.

In [3]:
from statistics import median
help(median)

Help on function median in module statistics:

median(data)
    Return the median (middle value) of numeric data.
    
    When the number of data points is odd, return the middle data point.
    When the number of data points is even, the median is interpolated by
    taking the average of the two middle values:
    
    >>> median([1, 3, 5])
    3
    >>> median([1, 3, 5, 7])
    4.0



In [4]:
median?

In [5]:
median??

### A brief overview of Markdown:

Q: What is markdown?

A: Markdown is a  simple markup language that uses plain-text syntax and functions pretty similarly to HTML. 


Q: What are some style things I can do with it?

# \# Headers


\*\*Bold text\*\* --> **Bold text**



\*Italics\* --> *Italics*


> \> Block quotes

Bulleted list:

- \- Item 1

- \- Item 2


Horizontal divider

Syntax: \-\-\-

---

\[Linked text\](link url) --> [Linked text](https://en.wikipedia.org/wiki/Python_(programming_language))




\`Code snippets\` --> `Code snippet`



\`\`\` 

Code block --> 

\`\`\`

```
{
Code block
}
```

Supports $\LaTeX$:

\frac{x^2+1}{3} --> $\frac{x^2+1}{3}$ 


    
</font>

Embeds image:

\!\[\]\(*Image Address*\) --> ![](https://upload.wikimedia.org/wikipedia/commons/3/38/Jupyter_logo.svg)

##### Markdown is used to style all the course materials, so it helps to have a little familiarity. 

As your Jupyter notebook can be exported in many formats, Markdown is a useful tool to learn for presenting your data analysis or machine learning model.

---

### Numpy Arrays

***Q: What are the `numpy` arrays? Why do we need them?***

`numpy` is one of the commonly used python modules/packages, which stands for numerical python. Numpy arrays are multidimensional arrays that are optimized for computing, especially for operations such as matrix multiplication.

To be able to use python modules, we first need to import them. 

In [1]:
import numpy as np
import pandas as pd

# The following two modules matplotlib and seaborn are for plots
import matplotlib.pyplot as plt
import seaborn as sns # Comment this if seaborn is not installed
%matplotlib inline

# The module re is for regular expressions
import re

### Pandas Dataframes

***Q: What are the*** `pandas` ***dataframes? Why do we need them? What is the crucial difference between numpy matrices and*** `pandas` ***dataframes?***

Pandas: an excellent tool to work with datasets

Dataframes: the central data structure of pandas library
- Evolved out of tables
- Most suitable for data manipulation tasks  

Pandas is built on top of numpy. The crucial difference between numpy matrices and pandas dataframes is that the columns in a Dataframe can be of different datatypes such as numerical, categorical, textual, etc.

First we load the [Titanic dataset from Kaggle](https://www.kaggle.com/c/titanic) stored in the `csv` file as a dataframe using [`read_csv`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) function.

In [15]:
path = 'data/titanic/'
df = pd.read_csv(path + 'train.csv')

If using Google Colab, the above will give you an error.

You will have to first download the `train.csv` file from the [Github repository](https://github.com/AashitaK/A-Hands-on-Workshop-series-in-Machine-Learning/tree/master/Session%201/titanic) then manually upload it to Colab using: 
```python
from google.colab import files
uploaded = files.upload()
```

Then, load the file into pandas dataframe using `read_csv`:
```python
df = pd.read_csv('train.csv')
```

In [9]:
df

As it turns out to be rather big dataset to display, we can comment the above cell by adding # in front of df and run it again to get rid of the output.

Next, let's check the numbers of rows and columns in the dataset.

In [10]:
df.shape

(891, 12)

So, the dataset consists of 891 rows and 12 columns.

We use `head()` function to peek into the first 5 rows (or any number of rows by using `head(n)`).

In [12]:
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

The [Titanic dataset](https://www.kaggle.com/c/titanic) that we will explore today is sourced from a [Kaggle](https://www.kaggle.com/) beginner-level competition.  

Goal of the competition: To apply the tools of machine learning to predict which passengers survived the Titanic tragedy.

[Description for the columns](https://www.kaggle.com/c/titanic/data) is as follows.  

|Variable|	Definition|	Key|   
|:---  |:--- |:---|
|PassengerId| Passenger ID |
|Survived| 	Survival|	0 = No, 1 = Yes |
|Pclass	|Ticket class|	1 = 1st, 2 = 2nd, 3 = 3rd|
|Sex	|Sex|	
|Age	|Age in years	|
|SibSp	|# of siblings / spouses aboard the Titanic	|
|Parch	|# of parents / children aboard the Titanic	|
|Ticket	|Ticket number	|
|Fare	|Passenger fare	|
|Cabin	|Cabin number	|
|Embarked	|Port of Embarkation	|C = Cherbourg, Q = Queenstown, S = Southampton|

Q: What are the features? 

Features are nothing but the variables in our model or the columns in our dataset. For example, `PClass`, `Age`, `Sex`, `Fare`, etc. are features for this particular dataset.

The final goal is to design a model to predict whether a passenger survives or not. 
* Which of the above features seem like important predictors? 
* How can you analyse the data in view of this objective?
    
Q: What is feature engineering?
* Detecting and handling missing values
* Encoding categorial features into numerical values
* Creating new features from the existing ones

---

### 1. Selecting rows and columns from the dataframe

How do we select a column from the dataframe? Say, we want to select the *Name* column from the dataframe. 

Remember, we used square brackets for indexing lists, strings and numpy arrays in Python, for example `A[0]`.

In [13]:
df['Name'].head()

0                              Braund, Mr. Owen Harris
1    Cumings, Mrs. John Bradley (Florence Briggs Th...
2                               Heikkinen, Miss. Laina
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                             Allen, Mr. William Henry
Name: Name, dtype: object

Since we do not want all the rows in the output, we have used `head()` function at the end.

How do we select **multiple columns**? Suppose we want to select the columns *Name, Sex* and *Age* from the dataframe. Hint: Use a list of columns inside the square brackets.

Name     Sex   Age
0                            Braund, Mr. Owen Harris    male  22.0
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0
2                             Heikkinen, Miss. Laina  female  26.0
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0
4                           Allen, Mr. William Henry    male  35.0

We can also select **rows** by putting a certain **condition on a column**. Say, we want only those rows for which the gender is *'female'*. 

In [20]:
df[df["Sex"] == "female"].head(3)

PassengerId  Survived  Pclass  \
1            2         1       1   
2            3         1       3   
3            4         1       1   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   

   Parch            Ticket     Fare Cabin Embarked  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S

Now, we want to retrieve only the female passengers traveling in the first class. Hint: Add another conditional `df['Pclass']==1` to the above code using the operator `&` and make sure to wrap the two conditionals with parenthesis `()`.

PassengerId  Survived  Pclass  \
1             2         1       1   
3             4         1       1   
11           12         1       1   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   

    Parch    Ticket     Fare Cabin Embarked  
1       0  PC 17599  71.2833   C85        C  
3       0    113803  53.1000  C123        S  
11      0    113783  26.5500  C103        S

We can also get the number of passengers using the shape method which gives us both the number of columns and the number of rows. Write the code to count the number of female passengers traveling in the first class. 

94

#### The `loc` and `iloc` methods
So far, we have seen how to retrieve:
* some select columns, or
* certain rows based on conditionals. 

What if we want to slice off a portion of the dataframe with some specific rows **and** columns? We use `.loc[]` or `.iloc[]` methods for this purpose. 
* `.iloc[]` method is primarily integer position based and gets rows/columns at particular positions in the index (so it **only takes integers** as index). 
* `loc[]` method is label based and gets rows/columns with particular labels from the index. Thus, it can be used to **put conditions on rows and retrieve select columns simultaneously**.

For example, we want to get the name and the survival information for all the adults above 70 years.

In [ ]:
df.loc[df['Age']>70, ['Name', 'Survived']]

Write the code to retrieve the **name, age and survival** information for all the **female passengers traveling in the first class**. 

Name   Age  Survived
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0         1
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0         1
11                            Bonnell, Miss. Elizabeth  58.0         1
31      Spencer, Mrs. William Augustus (Marie Eugenie)   NaN         1
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  49.0         1
..                                                 ...   ...       ...
856         Wick, Mrs. George Dennick (Mary Hitchcock)  45.0         1
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  48.0         1
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  47.0         1
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  56.0         1
887                       Graham, Miss. Margaret Edith  19.0         1

[94 rows x 3 columns]

The `iloc[]` method let us retrieve rows by passing sequence of indexes. For example, we can select the rows numbered 100th to 105th. The indexing works exactly like python lists and numpy arrays.

In [25]:
df.iloc[100:106]

PassengerId  Survived  Pclass                              Name     Sex  \
100          101         0       3           Petranec, Miss. Matilda  female   
101          102         0       3  Petroff, Mr. Pastcho ("Pentcho")    male   
102          103         0       1         White, Mr. Richard Frasar    male   
103          104         0       3        Johansson, Mr. Gustaf Joel    male   
104          105         0       3    Gustafsson, Mr. Anders Vilhelm    male   
105          106         0       3             Mionoff, Mr. Stoytcho    male   

      Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
100  28.0      0      0   349245   7.8958   NaN        S  
101   NaN      0      0   349215   7.8958   NaN        S  
102  21.0      0      1    35281  77.2875   D26        S  
103  33.0      0      0     7540   8.6542   NaN        S  
104  37.0      2      0  3101276   7.9250   NaN        S  
105  28.0      0      0   349207   7.8958   NaN        S

Write the code to retrieve every 100th row from the dataframe.

PassengerId  Survived  Pclass  \
0              1         0       3   
100          101         0       3   
200          201         0       3   
300          301         1       3   
400          401         1       3   
500          501         0       3   
600          601         1       2   
700          701         1       1   
800          801         0       2   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
100                            Petranec, Miss. Matilda  female  28.0      0   
200                     Vande Walle, Mr. Nestor Cyriel    male  28.0      0   
300           Kelly, Miss. Anna Katherine "Annie Kate"  female   NaN      0   
400                                 Niskanen, Mr. Juha    male  39.0      0   
500                                   Calic, Mr. Petar    male  17.0      0   
600  Jacobsohn, Mrs. Sidney Samuel (Amy Frances Chr...  female  24.0      2   
700  Astor, Mrs. John Jacob (Madeleine Talmadge Force)  female  18.0      1   
800                               Ponesell, Mr. Martin    male  34.0      0   

     Parch             Ticket      Fare    Cabin Embarked  
0        0          A/5 21171    7.2500      NaN        S  
100      0             349245    7.8958      NaN        S  
200      0             345770    9.5000      NaN        S  
300      0               9234    7.7500      NaN        Q  
400      0  STON/O 2. 3101289    7.9250      NaN        S  
500      0             315086    8.6625      NaN        S  
600      1             243847   27.0000      NaN        S  
700      0           PC 17757  227.5250  C62 C64        C  
800      0             250647   13.0000      NaN        S

Write the code to retrieve the last 10 rows from the dataframe using `iloc[]` method.

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

---

## Instructions for the exercise session:
- There are two exercise sections (section 2 and 3) below and they are alloted 45 min and 30 min respectively.
- The exercise involves new concepts not covered in the guided session above. Please feel free to ask questions and take help from me and/or your peers in the breakout room.
- The hints are provided for the each of the exercises. The built-in functions to be used for them are provided with a clickable link to the user manual. 
- The exercise sessions are time-bound and you are encouraged to work in groups to speed things up! 
- If you finished a section early, move on to the next one. There is an optional section 4 at the end.

### 2. Exploring the dataset (45 min)

Use [`describe`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) function for numerical features (columns) to get a brief overview of the statistics of the data.

Do the same as above for qualitative (non-numerical) features. Hint: Use `include=['O']` parameter in the [`describe`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html) function.

 Use the built-in pandas function to count the number of surviving and non-surviving passengers. Hint: Use [`value_counts()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html) on the column `df['Survived']`.

Below is a pie chart of the same using `matplotlib`:

In [ ]:
plt.axis('equal') 
plt.pie(df['Survived'].value_counts(), labels=('Died', "Survived"));

Below is a bar chart for the survival rate among male and female passengers using `seaborn` which we imported above as `sns`. Here is [Seaborn cheatsheet](https://s3.amazonaws.com/assets.datacamp.com/blog_assets/Python_Seaborn_Cheat_Sheet.pdf).

In [ ]:
sns.barplot(x = 'Sex', y = 'Survived', data = df);

Plot the survival rate among passengers in each ticket class.

We can also check the survival rate among both genders within the three ticket classes as follows.

In [ ]:
sns.barplot(x='Pclass', y='Survived', hue='Sex', data=df);

From the above chart, do you think that the gender affect the chance of survival for all the three ticket classes equally? Or does it seem like gender's effect is more pronounced for a certain ticket class passengers than others? We plot the  point estimates and confidence intervals for each sub-category to see it more clearly.

In [ ]:
sns.pointplot(x='Sex', y='Survived', hue='Pclass', data=df);

Notice the steeper slope for the second class.

It seems that gender and ticket class put together give more information about the survival chance than both of them separately. Please feel free to later explore other variables and combination of variables in depth in your own time.

How many children were on board? Hint: Use indexing on rows using conditional on the *Age* column and then the [`shape`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html) method to count the rows as seen above.

How many of the children on board survived? Hint: Add another conditional for the *Survived* column to the above code.

Use the functions [`isnull()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isnull.html) and [`sum()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sum.html) chained one after the other on the dataframe to find out the number of missing values in each column.

**Detecting missing values** is an important first step in Feature Engineering, that is preparing the features (independent variables) to use for building the machine learning models. 

The next step is **handling those missing values**. An option is to drop the rows or columns that have some or a lot of missing values, but that also means discarding relevant information. Another option is to fill them with something appropriate. 

Discuss:
1. What are the pros and cons of dropping the rows and/or columns with missing values?
    - Should you drop none, all or some of the columns with missing values for this dataset in view of building the predictive model? 
    - Ditto for rows with missing values.
3. If you consider filling the missing values, what are the possible options? 
    - Can you make use of other values in that column to fill the missing values? 
    - Can you make use of other values in that row as well as values in that column to fill the missing values 
4. Can the title in the name column be used for guessing a passengers' age based on the age values of other passengers with the same title?

What is the most common port of embarkment? Hint: Check the frequency (counts) of each value in the Embarked column using the built-in function [`value_counts()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html) as seen above. 

As we saw above, there are missing values in the column for *Embarked*. Fill them with the most commonly occuring value. Hint: Use [`fillna()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html).

Let us check whether the missing values for the *Embarked* column is indeed filled.

In [ ]:
df.isnull().sum()

If not, there are two options to fix this. One is to set `inplace` parameter in the `fillna()` function as `True` and another is to use assignment operator `=` as in `df = df.function()`. 

***Question***: Why is the `inplace` keyword False by default? This is true not just for `fillna()` but for most built-in functions in pandas. 

Answer: To facilitate method chaining or piping i.e. invoking multiple operations one after the other. For example, `df.isnull().sum()` used above. Chaining is more commonly used in pandas as compared to another programming style i.e. using nested function calls. Please read more [here](https://towardsdatascience.com/the-unreasonable-effectiveness-of-method-chaining-in-pandas-15c2109e3c69), if interested.

We should remove the *Cabin* column from the DataFrame -- too many values are missing. Hint: Use [`drop()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html) with appropriate value for the `axis` keyword. 

Let us check whether the column is indeed dropped. If not, modify the code above accordingly.

In [ ]:
df.head()

What is the age of the oldest person on board? 

Find all the passenger information for the oldest person on board. Hint: Use [`loc[]`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.loc.html) method with [`idxmax()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.idxmax.html) for the Age column.

---

### 3. Feature Engineering: Creating a new column for the titles of the passengers (30 min)

The real-world datasets many-a-times contain useful information in the textual format. Text mining is an important area of data science and one of the most powerful tool is [regular expressions](https://www.gnu.org/software/sed/manual/html_node/Regular-Expressions.html) that are not specific to python, but have much wider applications.

In this section, you are going to create a new feature for the titles of the passengers derived from their names using regular expressions. For that, let us first take a look at the passengers' names. 

In [ ]:
df.loc[:20, 'Name'].values

We notice one of the identifying characteristics of the titles above are that they end with a period. Regular expressions are very useful in the process of such data extraction and we will use them using the python module `re` to extract the titles from the *Name* column. We will first use regular expressions characters to construct a pattern and then use built-in function `findall` for pattern matching.

Some useful regular expression characters:
- `\w`: pattern must contain a word character, such as letters.
- `[ ]`: pattern must contain one of the characters inside the square brackets. If there is only one character inside the square brackets, for example `[.]`, then the pattern must contain it.

Let's try this.

In [ ]:
re.findall("\w\w[.]", 'Braund, Mr. Owen Harris')

It worked! It returned a list instead of the string, so we use indexing to get the first element of the list.

In [ ]:
re.findall("\w\w[.]", 'Braund, Mr. Owen Harris')[0]

Let us try it on another name:

In [ ]:
re.findall("\w\w[.]", 'Heikkinen, Miss. Laina')[0]

So, we want a pattern that automatically detects the length of the title and returns the entire title.

For regular expressions, \+ is added to a character/pattern to denote it is present one or more times. For example, `\w+` is used to denote one or more word characters. Fill in the regular expression in the below cell that will detect a period preceeded by one or more word characters.

The output should be `'Miss.'`

Summary: For pattern matching the titles using regular expressions:
- First we make sure it contains a period by using `[.]`. 
- Secondly, the period must be preceeded by word characters (one or more), so we use `\w+[.]`.

Write a function `get_title` that takes a name, extracts the title from it and returns the title.

Check that the function is working properly by running the following two cells.

In [ ]:
get_title('Futrelle, Mrs. Jacques Heath (Lily May Peel)')

The output should be `'Mrs.'`. Note: Make sure that the funtion returns a string and not a list. Please modify the above function accordingly.

In [ ]:
get_title('Simonius-Blumer, Col. Oberst Alfons')

The output should be `'Col.'`.

Create a new column named Title and extract titles from the Name column using the above function `get_title`. Hint: Use built-in [`map()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.map.html) function. The syntax is `df['New_column'] = df['Relevant_column'].map(function_name)`.

Let us peek into the dataframe.

In [ ]:
df.head()

List all the unique values for the titles along with their frequency. Hint: Use an inbuilt pandas function

Now, we want to replace the various spellings of the same title to a single one. Hint: Use the below dictionary with the [`replace`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.replace.html) function

`title_dictionary = {'Ms.': 'Miss.', 'Mlle.': 'Miss.', 
              'Dr.': 'Rare', 'Mme.': 'Mrs.', 
              'Major.': 'Rare', 'Lady.': 'Rare', 
              'Sir.': 'Rare', 'Col.': 'Rare', 
              'Capt.': 'Rare', 'Countess.': 'Rare', 
              'Jonkheer.': 'Rare', 'Dona.': 'Rare', 
              'Don.': 'Rare', 'Rev.': 'Rare'}`

List all the unique values for the titles along with their frequency to check that the titles are replaced properly.

What is the median age of passengers? Hint: Use the inbuilt function [`median`](http://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.median.html).

What is the median age of passengers with the title 'Miss.'? Hint: Use [`loc[]`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.loc.html) method for slicing off the select rows and the *Age* column.

What is the median age of passengers with the title 'Mrs.'?

Is there a noticeble difference in the median ages for the passengers with the above two titles? Should we take titles into account while filling the missing values for the *Age* column? If yes, how?

This is the end of the exercise session. If you finished early, move on to the next section for merging datasets.

---

### 4. Merging columns from different datasets (Optional) <a name="section4"></a>:
A simple illustration to merge two datasets using [`merge()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html).

In [ ]:
df1 = pd.DataFrame({'CourseCode': ['PHYS024', 'CSCI35', 'ENGR156'], 
                   'CourseName': ['Mechanics and Wave Motion', 
                                  'Computer Science for Insight',
                                 'Intro to Comm & Info Theory']})

df2 = pd.DataFrame({'Professor': ['Zachary Dodds', 'Vatche Sahakian', 
                                  'Timothy Tsai', 'Brian Shuve'],
                    'CourseCode': ['CSCI35', 'PHYS024',  'ENGR156', 'PHYS024']})

df1.head()

In [ ]:
df2.head()

In [ ]:
pd.merge(df2, df1)

Please refer to the documents [here](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) and [here](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html) to better grasp how and when to use `merge()` function. 

#### Merging dataframes from [Instacart Market Basket Analysis](https://www.kaggle.com/c/instacart-market-basket-analysis/overview) dataset:
Our next dataset is taken from a [competition](https://www.kaggle.com/c/instacart-market-basket-analysis/overview) hosted by Instacard on Kaggle platform. The dataset contains information about products in its grocery stores given in four files. Please see [here](https://www.kaggle.com/c/instacart-market-basket-analysis/data) for more information.
We load the four files into separate dataframes:   
`aisles.csv`   
`departments.csv`  
`products.csv`  
`order_products__train.csv`

In [ ]:
path = 'data/instacart-market-basket-analysis/'
dfa = pd.read_csv(path + 'aisles.csv')
dfd = pd.read_csv(path + 'departments.csv')
dfp = pd.read_csv(path + 'products.csv')
dfo = pd.read_csv(path + 'order_products__train.csv')

Familiarize yourself with the dataframes. Hint: Use [`head()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.head.html).  
Note: You might want to add the code cells. This can be done in two ways:
1. Using the keyboard shortcuts: 
    1. First press `Esc` key to enter the command mode.
    2. Add cell above using `A` or below using `B`.
    3. Exit by pressing `Enter/Return` key.
2. Using `Insert` tab in the top bar of the notebook and then `Insert Cell Above` or `Insert Cell Below`.


**Goal: Get a dataframe consisting of the name of the products along with their aisle names and department names for the order with `order_id` equal to 1.**  
This dataframe must have ***8 rows and only three columns:
```'product_name', 'aisle', 'department'```***

First slice out 8 rows from the order_products dataframe that corresponds with `order_id` equal to 1 and save it in a new dataframe `df`. Hint: Use conditional on indexing as seen above.

Let us have a peek into the dataframe `df`.

In [ ]:
df

The dataframe `df` has a column named `product_id`. So, we can merge it with the dataframe `dfp` to get `product_name`.

In [ ]:
dfp.head()

Merge the dataframes `df` and `dfp` using [`merge()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) and save the result back to `df`.

Let us check whether the new columns are added to the dataframe `df`.

In [ ]:
df

Now, we have new columns `aisle_id` and `department_id`. Let us first merge `dfa` with this dataframe to get the column `aisle`.

Let us check our dataframe `df` again. It should reflect the changes we attempted.

In [ ]:
df

Now it should have the column `aisle`. Let us now merge `dfd` with this dataframe to get the column `department`.

Let us check the dataframe `df` again. 

In [ ]:
df

Finally, we only select 3 columns: `'product_name', 'aisle', 'department'`. Hint: Use indexing with the list of columns as seen above.

Let us check the dataframe `df` again.

In [ ]:
df

Let us check the shape of th dataframe `df`.

In [ ]:
df.shape

The output should be `(8, 3)`. Please check your code above if you get something else.

#### Acknowledgment:
* [Titanic dataset from Kaggle](https://www.kaggle.com/c/titanic) dataset openly available in Kaggle is used in the exercises.
* [Instacart Market Basket Analysis](https://www.kaggle.com/c/instacart-market-basket-analysis/data) dataset openly available in Kaggle is used in the exercises.
